### DISCUSSION TITLE
Is it possible to restart a thread?

### DISCUSSION THREAD

**Discussion by marc-hg** [pybricks/support#1243] [Q&A] [answered]
Consider the following code:

```python
from pybricks.hubs import PrimeHub
from pybricks.tools import wait, multitask, run_task

hub = PrimeHub()

async def print_numbers():
    i = 0
    while True:
        hub.display.char(i)
        i += 1

print_numbers_task = print_numbers()

async def stop_it():
    await wait(1000)
    hello_world_task.close()

# This is the main program
async def main():
    print("Running multiple tasks at once!")
    await multitask(print_numbers, stop_it())

# Run the main program.
run_task(main())
```

My question is, do threads have a method to restart them? Right now **stop_it** stops **print_numbers** and I'd like to know if there's a way to instead of stop it, restart it

**Comment by laurensvalk** [comment]
```python
async def print_numbers():
    i = 0
    while True:
        hub.display.char(i)
        i += 1
```

This function does not have any yields or waits such as  `await wait(10)`, so it would block the rest of your program.

Calling `print_numbers()` again will create another object that can be awaited, starting from the beginning.

Depending on what you would like to do, maybe you can achieve it using the `race=True` option in `multitask` for a more natural program flow. For example:


```python
from pybricks.tools import multitask, run_task

async def some_function():
    print("Hello")
    await wait(1000)
    print("World!")

async def some_stop_condition():
    while my_value < 123: # some condition that suits your program
        await wait(10)

async def hello():
    await multitask(some_function(), some_stop_condition(), race=True)
    # rest of your code here. for example the same again to "restart" it
    await multitask(some_function(), some_stop_condition(), race=True)


run_task(hello())
```

**Reply by marc-hg** [reply]
Would it be possible to do this without restarting other tasks? I've simplified my example but in my real app I have a lot more tasks running in parallel in one giant multitask and I'd like to have control to turn them on/off without affecting others

**Reply by laurensvalk** [reply, marked as answer]
The `race` parameter only applies to the `multitask` statements where you use it. So you can modularize it to use the above technique locally instead of everywhere.

In the example above, the `hello()` function is just another task that can be used in another multitask statement that does not set the `race` parameter.

**Reply by marc-hg** [reply]
Thanks I didn't know multithreads could be stacked

**Reply by marc-hg** [reply]
Is it possible to stop a thread and then resume it? Like a .stop and .resume method for threads

**Reply by laurensvalk** [reply]
It's not built in, but you can probably build something yourself.

Note that the provided functionality is not unique to Pybricks. Everything works using so-called generators (a generic Python concept), so you can build rather advanced functionality if you like.
